### PR TITLE
Add complete swagger 3.0 security support for oauth2, apiKey and basic

### PR DESCRIPTION
**Pull request by franzwilhelm**
# Depends on https://github.com/astaxie/beego/pull/2640 to work

# In controller (can also be set globally in router)
### Definition
* **required**: securityPath
* **optional**: permission

*permission can be repeated x no of times*
### Example (related to oauth2 example)
```// @Security mySecurityPathNameOauth2 read:noMoreDogs write:noMoreDogs```
```json  
{
    "mySecurityPathNameOauth2": [
        "read:noMoreDogs",
        "write:noMoreDogs"
    ]
}

```

# oauth2
### Definition		
* **required**: securityPath	
* **required**: type
* **required**: authorizationUrl
* **required**: flow (implicit, password, application or accessCode)
* **required**: scopePath
* **required**: scopeDescription (quoted)
* **optional**: "your description of choice" (quoted)

*scopePath and description can be repeated as a pair x no of times*

### Example
```// @SecurityDefinition mySecurityPathNameOauth2 oauth2 http://www.ohmysweetoauthtest.com/dialog/oauth2 implicit write:noMoreDogs "write to this cool thingy" read:noMoreDogs "read from this cool thingy" "oauth2 is not complicated!"```
```json 
"mySecurityPathNameOauth2": {
     "type": "oauth2",
     "description": "oauth2 is not complicated!",
     "flow": "implicit",
     "authorizationUrl": "http://www.ohmysweetoauthtest.com/dialog/oauth2",
     "scopes": {
         "read:noMoreDogs": "read from this cool thingy",
         "write:noMoreDogs": "write to this cool thingy"
     }
}
```

image

# apiKey
### Definition
* **required**: securityPath
* **required**: type
* **required**: name
* **required**: in (header or query)
* **optional**: "your description of choice" (quoted)

### Example
```// @SecurityDefinition mySecurityPathNameApiKey apiKey Authorization header "I love auto-generated docs"```
```json
"mySecurityPathNameApiKey": {
      "type": "apiKey",
       "description": "I love auto-generated docs",
       "name": "Authorization",
       "in": "header"
}
```
image

# basic
### Definition
* **required**: securityPath
* **required**:  type
* **optional**:  "your description of choice" (quoted)

### Example
```// @SecurityDefinition mySecurityPathNameBasic  basic "I'm really just too basic."```
``` json
"mySecurityPathNameBasic": {
    "type": "basic",
    "description": "I'm really just too basic."
},
```
image

### CODE REVIEW

File: generate/swaggergen/g_docs.go [hunk 1]
```diff
@@ -184,6 +184,53 @@ func GenerateDocs(curpath string) {
 					rootapi.Schemes = strings.Split(strings.TrimSpace(s[len("@Schemes"):]), ",")
 				} else if strings.HasPrefix(s, "@Host") {
 					rootapi.Host = strings.TrimSpace(s[len("@Host"):])
+				} else if strings.HasPrefix(s, "@SecurityDefinition") {
+					if len(rootapi.SecurityDefinitions) == 0 {
+						rootapi.SecurityDefinitions = make(map[string]swagger.Security)
+					}
+					var out swagger.Security
+					p := getparams(strings.TrimSpace(s[len("@SecurityDefinition"):]))
+					out.Type = p[1]
```
Comment (astaxie) [inline generate/swaggergen/g_docs.go]: need add a len(p) before we access p[1], avoid panic here.

Comment (franzwilhelm) [inline generate/swaggergen/g_docs.go]: fixed here: cb52053d7f71d9c3fe6e40cd409e5ea5ffaa7c71

**Comment by franzwilhelm** [comment]
Now works with multiple `@SecurityDefinition` and `@Security`, in both router and controller

**Comment by astaxie** [comment]
@franzwilhelm I merge the pr in beego. Could you help to update the vendor? So the CI would passed. Thanks

**Comment by franzwilhelm** [comment]
@astaxie i don't know why the build is failing, have updated vendor

**Comment by amrfaissal** [comment]
You can create a new ignore list for `staticcheck` like `gosimple.ignore` and in it put the check code `SA01024`. Don't forget to feed the new ignore list to `staticcheck` in .travis.yml.

**Comment by franzwilhelm** [comment]
This came up with gosimple due to an update, so had to add new ignore for it as well. 
`utils/doc_value.go:8:9: should use *d instead of fmt.Sprint(*d) (S1026)`

**Comment by apremalal** [comment]
Appreciate all the hard work on this. Will this be reviewed merged to master in near future?

**Comment by dominikh** [comment]
SA1024 wasn't flagging that `if`, nor the comparison with `nil`. It was flagging the `strings.TrimLeft(c.Text, "//")`, which does not strip `//` from the beginning of a string. It strips all characters in the cut set. That is, `//// foo` would be changed to `foo`. And `//` is the same as `/`.

**Comment by Herts** [comment]
For those who are confused about how to set this:
Add the following line with securityDefinition.
`// @Security AUTH_NAME
`
For example, I've added this code:
`// @Security ApiKeyAuth`
`// @SecurityDefinition ApiKeyAuth apiKey Authorization header "Authorization token set in conf/admin.conf"`
`package routers`

**Pull request merged**